### PR TITLE
Fix background_pool_size not take effect and BackgroundProcessingPool::getThreadIds may misses some thread_ids.

### LINKED ISSUE #4684: Configuration profiles.default.background_pool_size not take effect.
### 1. Minimal reproduce step (Required)

Set configuration `profiles.default.background_pool_size: 8`.

### 2. What did you expect to see? (Required)

Create `BackgroundProcessingPool` with 8 threads.

### 3. What did you see instead (Required)

```
[2022/04/15 19:53:55.826 +08:00] [INFO] [<unknown>] ["BackgroundProcessingPool:Create BackgroundProcessingPool with 16 threads"] [thread_id=1]
```

### 4. What is your TiFlash version? (Required)

master

### PR DESCRIPTION
**Pull request by JinheLin**
### What problem does this PR solve?

Issue Number: close #4684 #4685

Problem Summary:

1. `Context::setDefaultProfiles` should be called before  `Context::initializeRateLimiter`. Because `Context::initializeRateLimiter` calls `Context::getBackgroundPool` and `Context::getBlockableBackgroundPool` inside to get background thread ids.
2. `BackgroundProcessingPool::getThreadIds` and `BackgroundProcessingPool::addThreadId` are executing concurrently. `BackgroundProcessingPool::getThreadIds` may be called before all background threads executing  ``BackgroundProcessingPool::addThreadId` completely.

### What is changed and how it works?

1. Adjust the initialization sequence of `Context`.
2. Add `absl::BlockingCounter` to count all background threads executing `BackgroundProcessingPool::addThreadId` completely.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

1. Set configuration `profiles.default.background_pool_size: 8`.
2. The expected log is found:

```
[2022/04/16 11:22:31.740 +08:00] [INFO] [BackgroundProcessingPool.cpp:90] ["BackgroundProcessingPool:Create BackgroundProcessingPool with 8 threads"] [thread_id=1]
[2022/04/16 11:22:31.763 +08:00] [INFO] [BackgroundProcessingPool.cpp:90] ["BackgroundProcessingPool:Create BackgroundProcessingPool with 8 threads"] [thread_id=1]
[2022/04/16 11:22:31.775 +08:00] [INFO] [RateLimiter.cpp:526] ["IORateLimiter:bg_thread_ids 16 => [9489, 9491, 9492, 9493, 9494, 9495, 9496, 9490, 9504, 9505, 9506, 9507, 9508, 9509, 9510, 9511]"] [thread_id=1]
```

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by Lloyd-Pottiger — APPROVED** [review]
LGTM

File: dbms/src/Server/Server.cpp [hunk 1]
```diff
@@ -1215,6 +1215,13 @@ int Server::main(const std::vector<std::string> & /*args*/)
     /// Init TiFlash metrics.
     global_context->initializeTiFlashMetrics();
 
+    /// Initialize users config reloader.
+    auto users_config_reloader = UserConfig::parseSettings(config(), config_path, global_context, log);
+
+    /// Load global settings from default_profile and system_profile.
+    global_context->setDefaultProfiles(config());
+    Settings & settings = global_context->getSettingsRef();
+
```
Comment (JaySon-Huang) [inline dbms/src/Server/Server.cpp:1224]: Call `global_context->getBackgroundPool()` and sent it to `initializeRateLimiter` explicitly?
And add more comments about the dependencies between config settings and init background pool

Comment (JinheLin) [reply dbms/src/Server/Server.cpp:1225]: Done.

**Review by JaySon-Huang — APPROVED** [review]

**Pull request merged**